x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleService.java

```diff
@@ -445,9 +449,34 @@ private void cancelJob() {
 
     @Override
     public void triggered(SchedulerEngine.Event event) {
-        if (event.jobName().equals(XPackField.INDEX_LIFECYCLE)) {
-            logger.trace("job triggered: {}, {}, {}", event.jobName(), event.scheduledTime(), event.triggeredTime());
-            triggerPolicies(clusterService.state(), false);
+        if (event.jobName().equals(XPackField.INDEX_LIFECYCLE) == false) {
+            assert false : "Expected scheduler event to be for ILM";
+            return;
+        }
+        logger.trace("job triggered: {}, {}, {}", event.jobName(), event.scheduledTime(), event.triggeredTime());
+
+        triggerPolicies(clusterService.state(), false);
+
+        // Check if we've received at least one cluster state update since the last periodic run.
+        // If not, we run all policies as if we just received a cluster state update. If there are any policies/indices currently in a step
+        // that is waiting for the next cluster state update (e.g. `wait-for-index-color`), they will get unstuck with this fallback.
+        if (didReceiveClusterStateSinceLastPeriodicRun.getAndSet(false) == false) {
+            // If a new cluster state came in while/after running the above check, or if ILM is still processing the last cluster state
+            // update that came in before the previous periodic run (i.e. when the processing thread is still in the queue),
+            // `lastSeenState` will be non-null, meaning we don't need to trigger the polices. That's why we only update `lastSeenState`
+            // if it was null before - to avoid redundant processing.
+            final var stateCurrentlyBeingProcessed = lastSeenState.compareAndExchange(null, clusterService.state());
+            if (stateCurrentlyBeingProcessed == null) {
+                logger.info("ILM didn't receive a new cluster state for [{}]. Running cluster state steps now", pollInterval);
+                processClusterState();
+            } else {
+                logger.warn(
+                    "ILM didn't receive a new cluster state for [{}] but it was still processing cluster state version [{}]",
```

Comment:
Should we clarify this to say something like "the poll interval should be increased" in the log message? In order to give a hint to users running on-prem.